lib/rex/logging/log_dispatcher.rb

```diff
@@ -134,8 +134,39 @@ def dlog(msg, src = 'core', level = 0)
   $dispatcher.log(LOG_DEBUG, src, level, msg)
 end
 
-def elog(msg, src = 'core', level = 0)
-  $dispatcher.log(LOG_ERROR, src, level, msg)
+# Logs errors in a standard format for each Log Level.
+#
+# @param msg [String] Contains message from the developer explaining why an error was encountered. Log Levels 0-3.
+#
+# @param src [String] USed to indicate where the error is originating from. Most commonly set to 'core' to ensure logs
+# are place in 'framework.log'.
+#
+# @param level [Integer] Sets the level an error should be logged at. Descriptions of each log level can be found in
+# lib/rex/logging.rb. Log Level will be changeD automatically to 0 if no +error+ is included.
+#
+# @param error [Exception] Exception of an error that needs to be logged. Mandatory in Log Levels 1-2. Optional in Log Level 3.
+#
+# (Eg Loop Iterations, Variables, Function Calls).
+#
+# @return [NilClass].
+def elog(msg, src = 'core', error:nil)
```

Comment:
I don't think we want to drop support for `level` here, that would be a backwards incompatible change

Reply:
yea I agree with that, correct me if I'm wrong but the global log level is what the user has said they want to see whereas the log level associated with the message is used to decide whether or not to show it to the user

What we have in here ends up just showing every message because every one is at the globally set level which is not what we want

Reply:
The two programmatic uses of `level` before this PR are two log sinks [stderr.rb](https://github.com/rapid7/metasploit-framework/blob/735dfa417ea02508861c7ea78b931dc1e5ecaf9a/lib/rex/logging/sinks/stderr.rb#L34) and [flatfile.rb](https://github.com/rapid7/metasploit-framework/blob/735dfa417ea02508861c7ea78b931dc1e5ecaf9a/lib/rex/logging/sinks/flatfile.rb#L44) (the `framework.log` file). Both sinks only pass the `level` value to be displayed in the log message, so I don't think it's value is used to decide what the user sees, it's more for people reading the logs at a later stage to be able to choose the verbosity levels of the messages they want to see.

My understanding of global log level was that it allowed users to decide how much detail would get added to their logfiles. Is this incorrect @adfoster-r7?

Reply:
Each individual call to the log functions will have their own corresponding log level.
If the right global log level set, the message will be passed on to the log sinks:

https://github.com/rapid7/metasploit-framework/blob/735dfa417ea02508861c7ea78b931dc1e5ecaf9a/lib/rex/logging/log_dispatcher.rb#L93

This function is being changed to further enrich the log messages if the global log level is set high enough.

The log message's log level, and the global log level, serve two separate purposes - and both are required